File: lib/geometry/Vector2D.cpp
#include <cmath>
#include <limits>
#include <string>

class Vector2D {
public:
  int x;
  int y;

  Vector2D(int x, int y) : x(x), y(y) {}
  ~Vector2D() {}
  Vector2D(const Vector2D &vector) : x(vector.x), y(vector.y) {}

  inline double norm() const {
    return std::sqrt(this->x * this->x + this->y * this->y);
  }

  inline Vector2D add(const Vector2D &vector) const {
    return Vector2D(this->x + vector.x, this->y + vector.y);
  }

  inline Vector2D subtract(const Vector2D &vector) const {
    return Vector2D(this->x - vector.x, this->y - vector.y);
  }

  inline long long dot(const Vector2D &vector) const {
    return static_cast<long long>(this->x) * vector.x +
           static_cast<long long>(this->y) * vector.y;
  }

  inline long long cross(const Vector2D &vector) const {
    return static_cast<long long>(this->x) * vector.y -
           static_cast<long long>(this->y) * vector.x;
  }

  inline Vector2D perpendicularVector() const {
    return Vector2D(-(this->y), this->x);
  }

  inline double angle() const {
    if (this->x == 0)
      return this->y > 0 ? std::numeric_limits<double>::infinity()
                         : std::numeric_limits<double>::lowest();
    return __builtin_atan2(this->y, this->x);
  }

  inline double angleBetween(const Vector2D &vector) const {
    double dotProduct = this->dot(vector);
    double normProduct = this->norm() * vector.norm();
    if (normProduct == 0)
      return (dotProduct > 0) ? std::numeric_limits<double>::infinity()
                              : std::numeric_limits<double>::lowest();
    return __builtin_acos(dotProduct / normProduct);
  }

  bool operator==(const Vector2D &vector) const {
    if (this == &vector)
      return true;
    return this->x == vector.x && this->y == vector.y;
  }

  inline std::string to_string() const {
    return std::to_string(this->x) + "i" + " + " + std::to_string(this->y) +
           "j";
  }
};

File: Library/java/src/geometry/Vector2D.java
package geometry;

public class Vector2D {

  public final int x;
  public final int y;

  public Vector2D(int x, int y) {
    this.x = x;
    this.y = y;
  }

  public double norm() {
    return Math.sqrt(this.x * this.x + this.y * this.y);
  }

  public Vector2D add(Vector2D vector) {
    return new Vector2D(this.x + vector.x, this.y + vector.y);
  }

  public Vector2D subtract(Vector2D vector) {
    return new Vector2D(this.x - vector.x, this.y - vector.y);
  }

  public long dot(Vector2D vector) {
    return (long) this.x * vector.x + (long) this.y * vector.y;
  }

  public long cross(Vector2D vector) {
    return (long) this.x * vector.y - (long) this.y * vector.x;
  }

  public Vector2D perpendicularVector() {
    return new Vector2D(-this.y, this.x);
  }

  public double angle() {
    if (this.x == 0)
      return this.y > 0 ? Double.POSITIVE_INFINITY : Double.NEGATIVE_INFINITY;
    return Math.atan2(this.y, this.x);
  }

  public double angleBetween(Vector2D vector) {
    double dotProduct = this.dot(vector);
    double normProduct = this.norm() * vector.norm();
    if (normProduct == 0)
      return (dotProduct > 0D) ? Double.POSITIVE_INFINITY : Double.NEGATIVE_INFINITY;
    return Math.acos(dotProduct / normProduct);
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj)
      return true;
    if (obj == null || !(obj instanceof Vector2D))
      return false;

    Vector2D vector = (Vector2D) obj;
    return this.x == vector.x && this.y == vector.y;
  }

  @Override
  public String toString() {
    return this.x + "i" + " + " + this.y + "j";
  }
}
